frontend/src/lib/utils.ts

```diff
@@ -203,3 +203,49 @@ export function labelVisibilityProtoValueToEnum(
       return LabelVisibilityOptions.Visible
   }
 }
+
+/**
+ * Tries to get an IFrame in which Streamlit app is embedded on Cloud deployments.
+ * It assumes iframe has title="streamlitApp", iterates over IFrames,
+ * and looks which IFrame contains div with stAppId value, otherwise returns first found iFrame or null.
+ */
+export function getIFrameEnclosingApp(
+  stAppId: string
+): HTMLIFrameElement | null {
+  if (!isInChildFrame()) {
+    return null
+  }
+  const qsStreamlitAppStr = 'iframe[title="streamlitApp"]'
+  let qs = window.document.querySelectorAll(
+    qsStreamlitAppStr
+  ) as NodeListOf<HTMLIFrameElement>
+  if (qs.length < 1 && window.parent) {
+    qs = window.parent.document.querySelectorAll(qsStreamlitAppStr)
+  }
+  for (let i = 0; i < qs.length; i++) {
+    const cd = qs[i].contentDocument
+    if (cd && cd.getElementsByClassName(`stAppId-${stAppId}`).length > 0) {
+      return qs[i]
+    }
+  }
+  let htmlCollection = window.document.getElementsByTagName(
+    "iframe"
+  ) as HTMLCollectionOf<HTMLIFrameElement>
+  if (htmlCollection.length < 1 && window.parent) {
+    htmlCollection = window.parent.document.getElementsByTagName("iframe")
+  }
+  if (htmlCollection.length > 0) {
+    return htmlCollection[0]
+  }
```

Comment:
We'll also want to have the ID check in the case that the iframe doesn't have the `streamlitApp` title.

I doubt there are many places where a site that isn't ours embeds >1 Streamlit app for now, but it'd be good to ensure this code works for that use-case + doesn't end up printing the wrong iframe in that case.

Since the two cases will end up looking so structurally similar, it might be best to factor the functionality out into a helper function that takes a query string that's passed either `'iframe[title="streamlitApp"]'` or `'iframe'`

Reply:
Heads up that this comment is relevant again now that I'm no longer panicking about strict cross-origin implications 😆